controllers/topology/internal/mergepatch/mergepatch.go

```diff
@@ -98,28 +109,41 @@ func NewHelper(original, modified client.Object, c client.Client, opts ...Helper
 }
 
 // filterPatch removes from the patch diffs not in the allowed paths;
-// it also return a flag indicating if the resulting patch has spec changes or not.
-func filterPatch(patch []byte, allowedPaths, ignorePaths []contract.Path) ([]byte, bool, error) {
-	// converts the patch into a Map
-	patchMap := make(map[string]interface{})
+// it also returns a flag indicating if the resulting patch has spec changes or not.
+func filterPatch(simplePatch, twoWayPatch []byte, allowedPaths, ignorePaths []contract.Path, authoritativePaths []contract.Path) ([]byte, bool, error) {
+	twoWayPatchMap := make(map[string]interface{})
+	if err := json.Unmarshal(twoWayPatch, &twoWayPatchMap); err != nil {
+		return nil, false, errors.Wrap(err, "failed to unmarshal two way merge patch")
+	}
+
+	// Enforce changes from simple patch when required.
+	// This will override instance specific fields for a subset of fields,
+	// e.g. machine template metadata changes should be reflected into generated objects without
+	// accounting for instance specific changes like we do for other maps into spec.
+	if len(authoritativePaths) > 0 {
```

Comment:
I wonder if what we're doing is more or less a subset of merging to mergepatches. If yes there is a func in the jsonpatch library which could fit: `jsonpatch.MergeMergePatches`.

> // MergeMergePatches merges two merge patches together, such that
> // applying this resulting merged merge patch to a document yields the same
> // as merging each merge patch to the document in succession.

That sounds very close to me what we want semantically: `patch(twoway-patch+authoritative-fields-patch)`

Reply:
Semantically yes, but in practice I'm not sure that modifying things in order to use this func will make the code more readable; also performance are kind of a concern.

Using MergeMergePatches will require to do both json -> map and map -> json twice (that is one Marshal more that now) + 1 additional call to jsonpatch.MergeMergePatches (which internally does a couple of other Marshal/Unmarshal, that with the current solution are not required).